File: archive/backend-node/src/services/post.service.ts
// Public blog CRUD and editorial-state management for admin content workflows.
import { BlogPost } from '../models/index.js';
import { BlogPostHydratedDocument } from '../models/blog-post.model.js';
import { AuthUser } from '../types/auth.js';
import { HTTP_STATUS } from '../constants/http-status.js';
import { ApiError } from '../utils/api-error.js';
import { toSlug } from '../utils/slug.js';
import { CreatePostInput, UpdatePostInput } from '../validations/post.validation.js';
import { emitPostCreated } from './notification-events.service.js';

function mapPost(post: BlogPostHydratedDocument) {
  const author = post.author as unknown;

  return {
    id: post._id.toString(),
    title: post.title,
    slug: post.slug,
    excerpt: post.excerpt,
    content: post.content,
    coverImage: post.coverImage,
    tags: post.tags,
    status: post.status,
    seoDescription: post.seoDescription,
    publishedAt: post.publishedAt,
    author:
      author && typeof author === 'object' && '_id' in (author as Record<string, unknown>)
        ? {
            id: String((author as { _id: unknown })._id),
            fullName:
              'fullName' in (author as Record<string, unknown>)
                ? String((author as { fullName?: unknown }).fullName ?? '')
                : undefined,
            email:
              'email' in (author as Record<string, unknown>)
                ? String((author as { email?: unknown }).email ?? '')
                : undefined,
          }
        : { id: String(post.author) },
  };
}

async function ensureUniquePostSlug(slug: string, excludePostId?: string) {
  const existing = await BlogPost.findOne({
    slug,
    ...(excludePostId ? { _id: { $ne: excludePostId } } : {}),
  });

  if (existing) {
    throw new ApiError(HTTP_STATUS.CONFLICT, 'Post slug already exists');
  }
}

async function findPostByIdOrThrow(postId: string) {
  const post = await BlogPost.findById(postId).populate('author', 'fullName email');

  if (!post) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Post not found');
  }

  return post;
}

export async function listPosts(user?: AuthUser) {
  const query =
    user?.role === 'admin'
      ? BlogPost.find()
      : BlogPost.find({ status: 'published' as const });
  const posts = (await query
    .populate('author', 'fullName email')
    .sort({ publishedAt: -1, createdAt: -1 })) as BlogPostHydratedDocument[];
  return posts.map((post) => mapPost(post));
}

export async function getPostBySlug(slug: string, user?: AuthUser) {
  const post = await BlogPost.findOne({ slug }).populate('author', 'fullName email');

  if (!post) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Post not found');
  }

  if (post.status !== 'published' && user?.role !== 'admin') {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Post not found');
  }

  return mapPost(post);
}

export async function createPost(input: CreatePostInput, user: AuthUser) {
  const slug = toSlug(input.slug ?? input.title);
  await ensureUniquePostSlug(slug);

  const post = await BlogPost.create({
    ...input,
    slug,
    author: user.userId,
    publishedAt: input.status === 'published' ? new Date() : undefined,
  });

  const created = await findPostByIdOrThrow(post._id.toString());
  emitPostCreated(user, {
    id: created._id.toString(),
    title: created.title,
    status: created.status,
  });
  return mapPost(created);
}

export async function updatePost(postId: string, input: UpdatePostInput) {
  const post = await findPostByIdOrThrow(postId);
  const nextSlug = input.slug ? toSlug(input.slug) : input.title ? toSlug(input.title) : undefined;

  if (nextSlug && nextSlug !== post.slug) {
    await ensureUniquePostSlug(nextSlug, postId);
  }

  const nextStatus = input.status ?? post.status;

  Object.assign(post, {
    ...(typeof input.title === 'string' ? { title: input.title } : {}),
    ...(typeof input.excerpt === 'string' ? { excerpt: input.excerpt } : {}),
    ...(typeof input.content === 'string' ? { content: input.content } : {}),
    ...(typeof input.coverImage === 'string' ? { coverImage: input.coverImage } : {}),
    ...(input.tags ? { tags: input.tags } : {}),
    ...(typeof input.seoDescription === 'string' ? { seoDescription: input.seoDescription } : {}),
    ...(typeof input.status === 'string' ? { status: input.status } : {}),
    ...(nextSlug ? { slug: nextSlug } : {}),
    publishedAt:
      nextStatus === 'published'
        ? post.publishedAt ?? new Date()
        : undefined,
  });

  await post.save();
  return mapPost(post);
}

export async function deletePost(postId: string) {
  const post = await BlogPost.findById(postId);

  if (!post) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'Post not found');
  }

  await post.deleteOne();
}

File: backend-spring/src/main/java/com/ivyts/backend/service/PostService.java
package com.ivyts.backend.service;

import com.ivyts.backend.common.exception.ApiException;
import com.ivyts.backend.domain.post.BlogPost;
import com.ivyts.backend.domain.post.BlogPostStatus;
import com.ivyts.backend.domain.user.User;
import com.ivyts.backend.domain.user.UserRole;
import com.ivyts.backend.notification.NotificationEventsService;
import com.ivyts.backend.security.AuthUser;
import com.ivyts.backend.service.poststore.PostStore;
import com.ivyts.backend.service.userstore.UserStore;
import com.ivyts.backend.util.SlugUtils;
import com.ivyts.backend.web.post.dto.CreatePostRequest;
import com.ivyts.backend.web.post.dto.UpdatePostRequest;
import java.time.Instant;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
public class PostService {

    private final PostStore postStore;
    private final UserStore userStore;
    private final NotificationEventsService notificationEventsService;

    public PostService(PostStore postStore, UserStore userStore, NotificationEventsService notificationEventsService) {
        this.postStore = postStore;
        this.userStore = userStore;
        this.notificationEventsService = notificationEventsService;
    }

    public List<Map<String, Object>> listPosts(AuthUser authUser) {
        return postStore.findAll().stream()
            .filter(post -> authUser != null && authUser.role() == UserRole.ADMIN || post.getStatus() == BlogPostStatus.PUBLISHED)
            .sorted(Comparator.comparing(BlogPost::getPublishedAt, Comparator.nullsLast(Comparator.reverseOrder()))
                .thenComparing(BlogPost::getCreatedAt, Comparator.nullsLast(Comparator.reverseOrder())))
            .map(this::toPostView)
            .toList();
    }

    public Map<String, Object> getPostBySlug(String slug, AuthUser authUser) {
        BlogPost post = postStore.findBySlug(slug)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Post not found"));

        if (post.getStatus() != BlogPostStatus.PUBLISHED && (authUser == null || authUser.role() != UserRole.ADMIN)) {
            throw new ApiException(HttpStatus.NOT_FOUND, "Post not found");
        }

        return toPostView(post);
    }

    public Map<String, Object> createPost(CreatePostRequest request, AuthUser authUser) {
        ensureAdmin(authUser);
        String slug = resolveUniqueSlug(SlugUtils.toSlug(request.slug() == null || request.slug().isBlank() ? request.title() : request.slug()), null);

        BlogPost post = new BlogPost();
        post.setAuthor(authUser.userId());
        post.setTitle(request.title().trim());
        post.setSlug(slug);
        post.setExcerpt(request.excerpt().trim());
        post.setContent(request.content().trim());
        post.setCoverImage(blankToNull(request.coverImage()));
        post.setTags(request.tags() == null ? List.of() : request.tags().stream().map(String::trim).filter(value -> !value.isBlank()).toList());
        post.setStatus(parseStatus(request.status()));
        post.setSeoDescription(blankToNull(request.seoDescription()));
        post.setPublishedAt(post.getStatus() == BlogPostStatus.PUBLISHED ? Instant.now() : null);
        BlogPost savedPost = postStore.save(post);
        notificationEventsService.emitPostCreated(
            authUser,
            savedPost.getId(),
            savedPost.getTitle(),
            savedPost.getStatus().name().toLowerCase(Locale.ROOT)
        );
        return toPostView(savedPost);
    }

    public Map<String, Object> updatePost(String postId, UpdatePostRequest request, AuthUser authUser) {
        ensureAdmin(authUser);
        BlogPost post = findPostOrThrow(postId);
        String nextSlug = post.getSlug();
        if (request.slug() != null && !request.slug().isBlank()) {
            nextSlug = resolveUniqueSlug(SlugUtils.toSlug(request.slug()), postId);
        }

        if (request.title() != null) post.setTitle(request.title().trim());
        if (request.excerpt() != null) post.setExcerpt(request.excerpt().trim());
        if (request.content() != null) post.setContent(request.content().trim());
        if (request.coverImage() != null) post.setCoverImage(blankToNull(request.coverImage()));
        if (request.tags() != null) post.setTags(request.tags().stream().map(String::trim).filter(value -> !value.isBlank()).toList());
        if (request.seoDescription() != null) post.setSeoDescription(blankToNull(request.seoDescription()));
        if (request.status() != null) {
            BlogPostStatus nextStatus = parseStatus(request.status());
            post.setStatus(nextStatus);
            post.setPublishedAt(nextStatus == BlogPostStatus.PUBLISHED ? (post.getPublishedAt() == null ? Instant.now() : post.getPublishedAt()) : null);
        }
        post.setSlug(nextSlug);

        return toPostView(postStore.save(post));
    }

    public Map<String, Object> deletePost(String postId, AuthUser authUser) {
        ensureAdmin(authUser);
        BlogPost post = findPostOrThrow(postId);
        postStore.delete(post);
        return Map.of();
    }

    private String resolveUniqueSlug(String slug, String excludePostId) {
        BlogPost existing = postStore.findBySlug(slug).orElse(null);
        if (existing != null && !existing.getId().equals(excludePostId)) {
            throw new ApiException(HttpStatus.CONFLICT, "Post slug already exists");
        }
        return slug;
    }

    private BlogPost findPostOrThrow(String postId) {
        return postStore.findById(postId).orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Post not found"));
    }

    private BlogPostStatus parseStatus(String value) {
        return BlogPostStatus.valueOf((value == null || value.isBlank() ? "draft" : value).trim().replace('-', '_').toUpperCase(Locale.ROOT));
    }

    private void ensureAdmin(AuthUser authUser) {
        if (authUser == null || authUser.role() != UserRole.ADMIN) {
            throw new ApiException(HttpStatus.FORBIDDEN, "Admin access is required");
        }
    }

    private String blankToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private Map<String, Object> toPostView(BlogPost post) {
        User author = userStore.findById(post.getAuthor()).orElse(null);
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("id", post.getId());
        data.put("title", post.getTitle());
        data.put("slug", post.getSlug());
        data.put("excerpt", post.getExcerpt());
        data.put("content", post.getContent());
        data.put("coverImage", post.getCoverImage());
        data.put("tags", post.getTags());
        data.put("status", post.getStatus().name().toLowerCase(Locale.ROOT));
        data.put("seoDescription", post.getSeoDescription());
        data.put("publishedAt", post.getPublishedAt());
        data.put("author", author == null ? Map.of("id", post.getAuthor()) : Map.of(
            "id", author.getId(),
            "fullName", author.getFullName(),
            "email", author.getEmail()
        ));
        return data;
    }
}
